micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jpa/HibernateMetrics.java

```diff
@@ -250,4 +264,97 @@ private SessionFactory unwrap(EntityManagerFactory entityManagerFactory) {
         }
     }
 
+    /**
+     * Register hibernate loadEvent listener - event trigger the metrics registration,
+     * only select queries are recorded in QueryStatistics {@link org.hibernate.stat.spi.StatisticsImplementor}
+     * @param meterRegistry meterRegistry to use
+     */
+    private void registerListeners(MeterRegistry meterRegistry) {
+        if (Objects.nonNull(sessionFactory) && sessionFactory instanceof SessionFactoryImplementor) {
+            EventListenerRegistry registry = ((SessionFactoryImplementor) sessionFactory).getServiceRegistry().getService(EventListenerRegistry.class);
+            MetricsEventHandler metricsEventHandler = new MetricsEventHandler(meterRegistry);
+            registry.appendListeners(EventType.POST_LOAD, metricsEventHandler);
+        }
+
+    }
+
+    class MetricsEventHandler implements PostLoadEventListener {
+
+        private final MeterRegistry meterRegistry;
+
+        MetricsEventHandler(MeterRegistry meterRegistry) {
+            this.meterRegistry = meterRegistry;
+        }
+
+        @Override
+        public void onPostLoad(PostLoadEvent event) {
+            registerQueryMetric(event.getSession().getFactory().getStatistics());
+        }
+
+        void registerQueryMetric(Statistics statistics) {
+            for (String query : statistics.getQueries()) {
+                QueryStatistics queryStatistics = statistics.getQueryStatistics(query);
```

Comment:
Is there any concern about this `queryStatistics` object being garbage collected? (since Micrometer favors weak references it won't keep the object around on its own)

At least that is how gauges work, not certain about `FunctionCounter`s

Reply:
The hibernate StatisticsImpl are keeping references to QueryStatistics in the Map, the entries aren't removed in any point of time. And since we are obtaining Statistics directly from SessionFactory which is created once during startup there no need to worry about their garbage collection, but it was good point didn't think about checkin' it before!

`public class StatisticsImpl implements StatisticsImplementor, Service {
      
private final ConcurrentMap<String, QueryStatisticsImpl> queryStatsMap = new ConcurrentHashMap();
}`